File: ip23/heateq_cg.m
%% Backward heat equation example from the sixth lecture.
close all
set(0,'DefaultLineLineWidth',2,'DefaultAxesFontSize',14,...
    'DefaultAxesFontWeight','bold');

% Initialize the parameters.
T = 0.1; % Time parameter.
n = 100; % Number of discretization points.
h = pi/n;% Step size.
x = linspace(h,pi-h,n-1); % Discretization grid.
F_true = double((x >= 1) .* (x <= 2))'; % Ground truth.
Nvec = 1:30; % Use 30 Fourier terms to compute the measurements at time t=T
[X,N] = meshgrid(x,Nvec);
Fcoefs = 2*(cos(N)-cos(2*N))./N/pi; % Fourier sine coefficients
U_true = dot(Fcoefs,exp(-N.^2*T).*sin(N.*X))'; % Analytic solution of the PDE at time t=T

% Construct the system matrix.
% First, use the stencil for the second order spatial derivative.
B = diag(ones(n-2,1),-1) - 2*eye(n-1) + diag(ones(n-2,1),1);
B = 1/h^2 * B;

% Create the forward operator for the temperature distribution at time T.
A = expm(T*B);

sigma = .01; % Standard deviation of normally distributed noise.
epsilon = sqrt(n-1)*sigma; % Noise level / Morozov discrepancy goal.

rng(123); % Reproducible experiments
U = U_true + sigma*randn(size(U_true));

%% We solve the normal equation using the conjugate gradient method
sol = zeros(n-1,1);   
residual = [];   % Residual of the normal equation
error = [];  % Residual of the original equation
k = 1;
r = A'*U;  % RHS of the normal equation
s = r;
residual(k) = norm(r);
error(k) = norm(U);
figure;
hold on;
h = legend;
while error(k) > epsilon
     hlp = A'*(A*s); % An auxiliary variable.
     alpha = residual(k)^2/(s'*hlp); % The line search in the direction s
     sol = sol + alpha*s; % New CG iterate
     r = r - alpha*hlp; % Update the residual
     residual(k+1) = norm(r); % Update the residual error
     beta =  residual(k+1)^2/residual(k)^2; % Compute the new search direction.
     s = r + beta*s;
     k = k+1;
     error(k) = norm(A*sol - U);
     plot(x,sol,'DisplayName',['k = ',num2str(k-1)]);
end

% Plot the results
plot(x,F_true,'r--','DisplayName','ground truth')
title('Evolution of CG iterates for A^TAF = A^TU');
hold off;

% Check that the Morozov discrepancy goal was achieved.
fprintf('CG reconstruction for A''AF = A''U\n');
disp(['Morozov discrepancy goal: ',num2str(epsilon)]);
disp(['Obtained discrepancy: ',num2str(norm(A*sol-U))]);

figure;
plot(x,sol,'b',x,F_true,'r--')
title(['CG reconstruction for A^TAF = A^TU (k = ',num2str(k-1),' chosen using Morozov)']);
legend('CG reconstruction','ground truth')

figure;
semilogy(error(2:end));
title('Residual ||AF_k-U||')

File: ip23/heateq_cg.py
# Backward heat equation example from the sixth lecture.
import numpy as np
from scipy import linalg
import matplotlib.pyplot as plt

# Initialize the parameters.
T = 0.1 # Time parameter.
n = 100 # Number of discretization points.
h = np.pi/n # Step size.
x = np.linspace(h,np.pi-h,n-1) # Discretization grid.
F_true = (x>=1) * (x<=2) * 1.0 # Ground truth.
Nvec = np.arange(1,31) # Use 30 Fourier terms to compute the measurements at time t=T
X,N = np.meshgrid(x,Nvec)
Fcoefs = 2*(np.cos(N)-np.cos(2*N))/N/np.pi
U_true = np.sum(np.multiply(Fcoefs,np.exp(-T*N**2)*np.sin(N*X)),axis=0) # Analytic solution of the PDE at time t=T

# Construct the system matrix.
# First, use the stencil for the second order spatial derivative.
B = np.diag(np.ones(n-2),-1) - 2*np.diag(np.ones(n-1)) + np.diag(np.ones(n-2),1)
B = B/h**2

# Create the forward operator for the temperature distribution at time T.
A = linalg.expm(T*B)

sigma = .01 # Standard deviation of normally distributed noise.
epsilon = np.sqrt((n-1))*sigma # Noise level / Morozov discrepancy goal.

np.random.seed(321); # Reproducible experiments
U = U_true + sigma * np.random.normal(size=U_true.shape)

# We solve the normal equation using the conjugate gradient method
sol = np.zeros(n-1)
residual = []
error = []
k = 0
r = A.T@U # RHS of the normal equation
s = r
residual.append(np.linalg.norm(r))
error.append(np.linalg.norm(U))
fig, ax = plt.subplots()

while error[k] > epsilon:
    hlp = A.T @ (A@s) # An auxiliary variable
    alpha = residual[k]**2/(np.dot(s,hlp)) # The line search in the direction s
    sol = sol + alpha*s # New CG iterate
    r = r-alpha*hlp # % Update the residual
    residual.append(np.linalg.norm(r)) # Update the residual error
    beta = residual[k+1]**2/residual[k]**2 # Compute the new search direction
    s = r + beta*s
    k = k+1
    error.append(np.linalg.norm(A@sol-U))
    ax.plot(x,sol,linewidth=2,label='k = ' + str(k))

# Check that the Morozov discrepancy goal was achieved.
print('CG reconstruction for A\'AF = A\'U')
print('Morozov discrepancy goal: ' + str(epsilon))
print('Obtained discrepancy: ' + str(np.linalg.norm(A@sol-U)))

# Plot the results
ax.plot(x,F_true,'--r',linewidth=2,label='ground truth')
ax.set_title('Evolution of CG iterates for $A^TAF = A^TU$', fontsize=15)
plt.legend(loc = 'upper right')
plt.show()

fig, ax = plt.subplots()
ax.plot(x,sol,'-b',linewidth=2,label='CG reconstruction')
ax.plot(x,F_true,'--r',linewidth=2,label='ground truth')
ax.set_title('CG reconstruction for $A^TAF = A^TU$\n(k = ' + str(k) + ' chosen using Morozov)',fontsize=15)
plt.legend(loc = 'upper right')
plt.show()

fig, ax = plt.subplots()
ax.semilogy(np.arange(1,k+1),error[1:])
ax.set_title('Residual $||AF_k-U||$',fontsize=15)
plt.show()